xarray/core/common.py

```diff
@@ -22,12 +22,16 @@ class ImplementsArrayReduce(object):
     def _reduce_method(cls, func, include_skipna, numeric_only):
         if include_skipna:
             def wrapped_func(self, dim=None, axis=None, skipna=None,
-                             keep_attrs=_set_keep_attrs(False), **kwargs):
+                             keep_attrs=None, **kwargs):
+                if keep_attrs is None:
+                    keep_attrs = _get_keep_attrs(default=False)
```

Comment:
Note: you probably need far fewer of these checks -- only functions that actually implement the `keep_attrs` logic need it, not functions that pass on `keep_attrs` to other functions with the same default value.

Reply:
I think I know what you mean, but in this case if I remove that check and just pass `keep_attrs=None` down into the `self.reduce` method, then some unit tests will fail (specifically `TestDataArray.test_reduce_keep_attrs`, `TestDataset.test_apply`, `TestVariable.test_reduce_keep_attrs`).

Reply:
This should be fixed. 

For e.g. the `DataArray` methods below usually end up calling `Variable` or `Dataset` methods so you should only have to change the default to None in `dataarray.py`

Reply:
Ah I got it - changed in most recent commit